x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsConfigTests.java

```diff
@@ -88,6 +90,49 @@ protected DataFrameAnalyticsConfig createTestInstance() {
         return createRandom(randomValidId(), lenient);
     }
 
+    @Override
+    protected List<Version> bwcVersions() {
+        return Arrays.asList(Version.V_7_0_0,
+            Version.V_7_1_1,
+            Version.V_7_2_1,
+            Version.V_7_3_2,
+            Version.V_7_4_2,
+            Version.V_7_5_3,
+            Version.V_7_6_0,
+            Version.V_7_7_0);
+    }
+
+    @Override
+    protected DataFrameAnalyticsConfig mutateInstanceForVersion(DataFrameAnalyticsConfig instance, Version version) {
+        DataFrameAnalyticsConfig.Builder builder = new DataFrameAnalyticsConfig.Builder(instance);
+        if (version.before(Version.V_7_6_0)) {
+            if (instance.getAnalysis() instanceof Regression) {
+                // TODO regression and classification serialize a randomized value for the random seed
+            }
+            builder.setSource(new DataFrameAnalyticsSource.Builder(instance.getSource())
+                .setSourceFiltering(null)
+                .build());
+        }
+        if (version.before(Version.V_7_5_0)) {
+            builder.setAllowLazyStart(false);
+            if (instance.getAnalysis() instanceof OutlierDetection) {
```

Comment:
This bit seems like it belongs in the tests for `OutlierDetection`. However, we'd need to access that mutation from here too to make this test work. We could reuse the `mutateInstanceForVersion` method of other tests here (by extracting the logic in a static method, similarly to our `createRandom` methods). However, I don't see how we'd get rid of the `instanceof` here.

One thing this makes me wonder is if it'd be better to design this in a way where we have a separate test for each BWC change we're testing. Say we have for example a private test method that takes some params and we can call it from different tests where we isolate the change we're testing. That would help avoiding `mutateInstanceForVersion` become too complex.

I like the idea of having these tests a lot, don't take me wrong! I think it's great! Just pondering a few ideas on the details.

Reply:
> One thing this makes me wonder is if it'd be better to design this in a way where we have a separate test for each BWC change we're testing.

We could potentially do this. But I also think that this type of separation could easily be done within the `mutateInstanceForVersion` by having separate function calls. 

My goal is to implicitly force the users of the scaffold to handle all BWC versions. Whether or not they want to. Having users write their own individual bwc tests for specific versions sort of defeats this.

Reply:
I agree though, that we should attempt to have static methods for us to reuse else where if an object has multiple sub-objects that have BWC tests. I will refactor a bit to handle that :D